File: ios/Ripple/Core/Backup.swift
import Foundation
import CryptoKit

/// Identity backup & restore (ROADMAP Phase 0.3, app-layer — no wire changes).
///
/// The exportable identity material is the 32-byte P-256 private scalar plus the
/// 65-byte uncompressed public key wire form (97 bytes). That payload is encrypted with
/// a passphrase-derived key (PBKDF2-HMAC-SHA256 → AES-256-GCM) and base64url'd into a
/// single text blob designed for offline transport: a QR screen on one phone, pasted
/// text over any channel, even a photo of a screen. Restoring on a fresh install
/// re-derives the *same* node id and key, so the device keeps decrypting its old
/// messages and peers keep verifying its signatures.
///
/// **A leaked passphrase means a lost identity**: whoever holds blob + passphrase *is*
/// this node — Ripple v1 has no revocation (that lands with key rotation in Phase 2.2).
/// See docs/BACKUP.md for the format and the security notes.
///
/// Pure Foundation/CryptoKit on purpose (like `Pairing`): the byte-level vectors are
/// mirrored verbatim in `BackupTests.swift` / `BackupTest.kt`, so a blob written on
/// either platform decodes on the other.
enum Backup {
    static let prefix = "RIPPLE-BKP"
    static let blobVersion = "v1"
    static let saltSize = 16
    static let nonceSize = 12
    static let tagSize = 16
    static let keySize = 32
    static let scalarSize = 32
    static let wireSize = 65
    static let payloadSize = scalarSize + wireSize                    // 97
    static let rawSize = saltSize + nonceSize + payloadSize + tagSize // 141

    /// PBKDF2 work factor. Chosen so restore takes well under a second on low-end
    /// phones while keeping online guessing of a decent passphrase infeasible; the
    /// blob is offline, so this guards the offline copy too — see docs/BACKUP.md §4.
    static let pbkdf2Iterations = 150_000

    /// Domain separation: this key is for Ripple backup blobs and nothing else.
    static let aad = Data("ripple/backup/v1".utf8)

    enum Failure { case format, base64, length, passphrase }

    struct ReadResult {
        let payload: Data?
        let failure: Failure?
        var ok: Bool { payload != nil }
    }

    /// Produce the canonical text blob: `RIPPLE-BKP:v1:<base64url(salt‖nonce‖ct‖tag)>`.
    /// Random salt/nonce — the same inputs give a different output each call.
    static func createBlob(scalar: Data, publicKeyWire: Data, passphrase: String) -> String {
        createBlob(scalar: scalar, publicKeyWire: publicKeyWire, passphrase: passphrase,
                   salt: Crypto.randomBytes(saltSize), nonce: Crypto.randomBytes(nonceSize))
    }

    /// Deterministic variant (fixed salt+nonce) — used by the shared cross-platform test vectors.
    static func createBlob(scalar: Data, publicKeyWire: Data, passphrase: String, salt: Data, nonce: Data) -> String {
        precondition(scalar.count == scalarSize && publicKeyWire.count == wireSize, "bad identity material sizes")
        precondition(salt.count == saltSize && nonce.count == nonceSize, "bad salt/nonce sizes")
        let key = SymmetricKey(data: deriveKey(passphrase: passphrase, salt: salt))
        // The only throw here is an invalid nonce size, which the precondition above rules out
        // (programmer-error invariant).
        let sealed = try! AES.GCM.seal(scalar + publicKeyWire, using: key,
                                       nonce: AES.GCM.Nonce(data: nonce), authenticating: aad)
        return "\(prefix):\(blobVersion):" + encodeBase64Url(salt + nonce + sealed.ciphertext + sealed.tag)
    }

    /// Parse and decrypt a blob. Whitespace/newlines anywhere in `blob` are tolerated
    /// (paste from messages or emails), and a standard base64 alphabet is accepted even
    /// though canonical form is URL-safe. A GCM auth failure is reported as `.passphrase`
    /// — it is indistinguishable from (and the expected answer to) a wrong passphrase;
    /// tampering never yields plaintext.
    static func readBlob(_ blob: String, passphrase: String) -> ReadResult {
        let scalars = blob.unicodeScalars.filter { !CharacterSet.whitespacesAndNewlines.contains($0) }
        let text = String(String.UnicodeScalarView(scalars))
        let parts = text.split(separator: ":", maxSplits: 2, omittingEmptySubsequences: false)
        guard parts.count == 3,
              parts[0].caseInsensitiveCompare(prefix) == .orderedSame,
              parts[1].caseInsensitiveCompare(blobVersion) == .orderedSame else { return ReadResult(payload: nil, failure: .format) }
        guard let raw = decodeBase64Url(String(parts[2])) else { return ReadResult(payload: nil, failure: .base64) }
        guard raw.count == rawSize else { return ReadResult(payload: nil, failure: .length) }
        let salt = raw.subdata(in: 0..<saltSize)
        let combined = raw.subdata(in: saltSize..<raw.count) // nonce‖ct‖tag
        let key = SymmetricKey(data: deriveKey(passphrase: passphrase, salt: salt))
        guard let box = try? AES.GCM.SealedBox(combined: combined),
              let payload = try? AES.GCM.open(box, using: key, authenticating: aad) else {
            return ReadResult(payload: nil, failure: .passphrase)
        }
        return ReadResult(payload: payload, failure: nil)
    }

    // MARK: payload accessors

    static func scalar(of payload: Data) -> Data { payload.subdata(in: 0..<scalarSize) }
    static func publicKey(of payload: Data) -> Data { payload.subdata(in: scalarSize..<payloadSize) }

    /// Node id (lowercase hex) the restored payload re-derives — for UI confirmation.
    static func nodeIdHex(of payload: Data) -> String { NodeId.fromPublicKey(publicKey(of: payload)).hex }

    /// PBKDF2-HMAC-SHA256 over the passphrase's raw UTF-8 bytes (Android's
    /// `PBEKeySpec` hashes the same bytes; so type the passphrase identically on both
    /// platforms — docs/BACKUP.md §4). CryptoKit has no PBKDF2, so this is the standard
    /// loop; dkLen (32) fits one HMAC block. 150k iterations ≈ tens of milliseconds,
    /// matching the JVM's SecretKeyFactory.
    static func deriveKey(passphrase: String, salt: Data) -> Data {
        let password = SymmetricKey(data: Data(passphrase.utf8))
        let block = salt + Data([0, 0, 0, 1])
        var u = Array(HMAC<SHA256>.authenticationCode(for: block, using: password))
        var t = u
        for _ in 1..<pbkdf2Iterations {
            u = Array(HMAC<SHA256>.authenticationCode(for: Data(u), using: password))
            for i in 0..<32 { t[i] ^= u[i] }
        }
        return Data(t)
    }

    // MARK: base64url (RFC 4648 §5), tolerant of the standard alphabet and padding

    static func encodeBase64Url(_ data: Data) -> String {
        data.base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }

    static func decodeBase64Url(_ value: String) -> Data? {
        let std = value
            .replacingOccurrences(of: "-", with: "+")
            .replacingOccurrences(of: "_", with: "/")
            .replacingOccurrences(of: "=", with: "")
        if std.count % 4 == 1 { return nil }
        let padded = std + String(repeating: "=", count: (4 - std.count % 4) % 4)
        return Data(base64Encoded: padded)
    }
}

File: android/app/src/main/java/app/ripple/mesh/core/Backup.kt
package app.ripple.mesh.core

import javax.crypto.Cipher
import javax.crypto.SecretKeyFactory
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.PBEKeySpec
import javax.crypto.spec.SecretKeySpec

/**
 * Identity backup & restore (ROADMAP Phase 0.3, app-layer — no wire changes).
 *
 * The exportable identity material is the 32-byte P-256 private scalar plus the
 * 65-byte uncompressed public key wire form (97 bytes). That payload is encrypted with
 * a passphrase-derived key (PBKDF2-HMAC-SHA256 → AES-256-GCM) and base64url'd into a
 * single text blob designed for offline transport: a QR screen on one phone, pasted
 * text over any channel, even a photo of the screen. Restoring on a fresh install
 * re-derives the *same* node id and key, so the device keeps decrypting its old
 * messages and peers keep verifying its signatures.
 *
 * **A leaked passphrase means a lost identity**: whoever holds blob + passphrase *is*
 * this node — Ripple v1 has no revocation (that lands with key rotation in Phase 2.2).
 * See docs/BACKUP.md for the format and the security notes.
 *
 * Pure JVM on purpose (like [Pairing]): the byte-level vectors are mirrored verbatim
 * in `BackupTest.kt` / `BackupTests.swift`, so a blob written on one platform decodes
 * on the other.
 */
object Backup {
    private const val PREFIX = "RIPPLE-BKP"
    private const val BLOB_VERSION = "v1"
    const val SALT_SIZE = 16
    const val NONCE_SIZE = 12
    const val TAG_SIZE = 16
    const val KEY_SIZE = 32
    const val SCALAR_SIZE = 32
    const val WIRE_SIZE = 65
    const val PAYLOAD_SIZE = SCALAR_SIZE + WIRE_SIZE           // 97
    const val RAW_SIZE = SALT_SIZE + NONCE_SIZE + PAYLOAD_SIZE + TAG_SIZE // 141

    /**
     * PBKDF2 work factor. Chosen so restore takes well under a second on low-end
     * phones while keeping online guessing of a decent passphrase infeasible; the
     * blob is offline, so this only guards the offline copy too — see docs/BACKUP.md §4.
     */
    const val PBKDF2_ITERATIONS = 150_000

    /** Domain separation: this key is for Ripple backup blobs and nothing else. */
    private val AAD: ByteArray = "ripple/backup/v1".toByteArray(Charsets.UTF_8)

    enum class Failure { FORMAT, BASE64, LENGTH, PASSPHRASE }

    /** [payload] is non-null exactly when [failure] is null. */
    class ReadResult(val payload: ByteArray?, val failure: Failure?) {
        val ok: Boolean get() = payload != null
    }

    /**
     * Produce the canonical text blob: `RIPPLE-BKP:v1:<base64url(salt‖nonce‖ct‖tag)>`.
     * Random salt/nonce; same inputs, different output each call.
     */
    fun createBlob(scalar: ByteArray, publicKeyWire: ByteArray, passphrase: String): String =
        createBlob(scalar, publicKeyWire, passphrase, Crypto.randomBytes(SALT_SIZE), Crypto.randomBytes(NONCE_SIZE))

    /** Deterministic variant (fixed salt+nonce) — used by the shared cross-platform test vectors. */
    fun createBlob(scalar: ByteArray, publicKeyWire: ByteArray, passphrase: String, salt: ByteArray, nonce: ByteArray): String {
        require(scalar.size == SCALAR_SIZE && publicKeyWire.size == WIRE_SIZE) { "bad identity material sizes" }
        require(salt.size == SALT_SIZE && nonce.size == NONCE_SIZE) { "bad salt/nonce sizes" }
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.ENCRYPT_MODE, SecretKeySpec(deriveKey(passphrase, salt), "AES"), GCMParameterSpec(TAG_SIZE * 8, nonce))
        cipher.updateAAD(AAD)
        val ctAndTag = cipher.doFinal(scalar + publicKeyWire)
        return PREFIX + ":" + BLOB_VERSION + ":" + encodeBase64Url(salt + nonce + ctAndTag)
    }

    /**
     * Parse and decrypt a blob. Whitespace/newlines anywhere in [blob] are tolerated
     * (paste from messages or emails), and a standard-base64 alphabet is accepted
     * even though canonical form is URL-safe. A GCM auth failure is reported as
     * [Failure.PASSPHRASE] — it is indistinguishable from (and the expected answer to)
     * a wrong passphrase; tampering never yields plaintext.
     */
    fun readBlob(blob: String, passphrase: String): ReadResult {
        val text = blob.filterNot { it.isWhitespace() }
        val parts = text.split(':', limit = 3)
        if (parts.size != 3 || !parts[0].equals(PREFIX, ignoreCase = true) || !parts[1].equals(BLOB_VERSION, ignoreCase = true)) {
            return ReadResult(null, Failure.FORMAT)
        }
        val raw = decodeBase64Url(parts[2]) ?: return ReadResult(null, Failure.BASE64)
        if (raw.size != RAW_SIZE) return ReadResult(null, Failure.LENGTH)
        return runCatching {
            val cipher = Cipher.getInstance("AES/GCM/NoPadding")
            cipher.init(
                Cipher.DECRYPT_MODE,
                SecretKeySpec(deriveKey(passphrase, raw.copyOfRange(0, SALT_SIZE)), "AES"),
                GCMParameterSpec(TAG_SIZE * 8, raw.copyOfRange(SALT_SIZE, SALT_SIZE + NONCE_SIZE)),
            )
            cipher.updateAAD(AAD)
            ReadResult(cipher.doFinal(raw.copyOfRange(SALT_SIZE + NONCE_SIZE, raw.size)), null)
        }.getOrDefault(ReadResult(null, Failure.PASSPHRASE))
    }

    // ---- payload accessors ----------------------------------------------------------

    fun scalarOf(payload: ByteArray): ByteArray = payload.copyOfRange(0, SCALAR_SIZE)
    fun publicKeyOf(payload: ByteArray): ByteArray = payload.copyOfRange(SCALAR_SIZE, PAYLOAD_SIZE)

    /** Node id (lowercase hex) the restored payload re-derives — for UI confirmation. */
    fun nodeIdOf(payload: ByteArray): String = NodeId.fromPublicKey(publicKeyOf(payload)).hex

    /**
     * PBKDF2-HMAC-SHA256 over the passphrase's UTF-8 bytes (the JVM applies the
     * password chars as-is; iOS mirrors this by hashing the raw UTF-8 bytes, so the
     * passphrase must be typed identically on both platforms — see docs/BACKUP.md §4).
     */
    internal fun deriveKey(passphrase: String, salt: ByteArray): ByteArray {
        val spec = PBEKeySpec(passphrase.toCharArray(), salt, PBKDF2_ITERATIONS, KEY_SIZE * 8)
        return SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256").generateSecret(spec).encoded
    }

    // ---- base64url (RFC 4648 §5), tolerant of the standard alphabet and padding ------

    private fun encodeBase64Url(bytes: ByteArray): String =
        java.util.Base64.getUrlEncoder().withoutPadding().encodeToString(bytes)

    private fun decodeBase64Url(value: String): ByteArray? {
        val std = value.replace('-', '+').replace('_', '/').replace("=", "")
        if (std.length % 4 == 1) return null
        val padded = std + "=".repeat((4 - std.length % 4) % 4)
        return runCatching { java.util.Base64.getDecoder().decode(padded) }.getOrNull()
    }
}
